lib/elixir/lib/kernel/parallel_compiler.ex

```diff
@@ -226,26 +229,53 @@ defmodule Kernel.ParallelCompiler do
         end
         spawn_compilers(state)
 
-      {:DOWN, _down_ref, :process, down_pid, {:shutdown, file}} ->
+      {:file_compiled, child_pid, file, :ok} ->
+        wait_for_down(queued, child_pid)
+
         if callback = Keyword.get(options, :each_file) do
           callback.(file)
         end
 
-        cancel_waiting_timer(queued, down_pid)
+        cancel_waiting_timer(queued, child_pid)
 
         # Sometimes we may have spurious entries in the waiting
         # list because someone invoked try/rescue UndefinedFunctionError
-        new_entries = List.delete(entries, down_pid)
-        new_queued  = List.keydelete(queued, down_pid, 0)
-        new_waiting = List.keydelete(waiting, down_pid, 1)
+        new_entries = List.delete(entries, child_pid)
+        new_queued  = List.keydelete(queued, child_pid, 0)
+        new_waiting = List.keydelete(waiting, child_pid, 1)
         spawn_compilers(%{state | entries: new_entries, waiting: new_waiting, queued: new_queued})
 
-      {:DOWN, down_ref, :process, _down_pid, reason} ->
-        handle_failure(down_ref, reason, queued)
+      {:file_compiled, child_pid, file, {kind, reason, stack}} ->
+        wait_for_down(queued, child_pid)
+        print_error(file, kind, reason, stack)
+        terminate(queued)
+
+      {:DOWN, ref, :process, _pid, reason} ->
+        handle_down(queued, ref, reason)
         wait_for_messages(state)
     end
   end
 
+  defp wait_for_down(queued, pid) do
+    receive do
+      {:DOWN, ref, :process, ^pid, reason} -> handle_down(queued, ref, reason)
+    end
+  end
+
+  defp handle_down(_queued, _ref, :normal) do
+    :ok
+  end
+
+  defp handle_down(queued, ref, reason) do
+    case List.keyfind(queued, ref, 1) do
+      {_child, ^ref, file, _timer_ref} ->
+        print_error(file, :exit, reason, [])
```

Comment:
Ugh the semantics here are awkward because we are pretending the error occurred in the the current process. I think the code is correct, and not my comment.